## What is Sentiment Analysis?
Sentiment Analysis is the process of ‘computationally’ determining whether a piece of writing is positive, negative or neutral. It’s also known as opinion mining, deriving the opinion or attitude of a speaker.

## Popular applications of Sentiment Analysis
- Movies: To understand people's thoughts on the movies.
- Business: Helps a business understand their customers' feelings towards their products or brand; develop strategies for compaigns or product launches.
- Politics: To interpret people's views on the statements and actions at the government level. Also, could be used to predict election results!
- Public Actions: To monitor and analyse social phenomena, for the spotting of potentially dangerous situations and determining the general mood of the blogosphere.

### Python and pre-requisites

I have launched Jupytor notebooks which is included when downloading Anaconda. I like Jupytor notebooks as it is easier to publish and separate notes from code.

Prior to running code in Python, there are a few packages to be installed:
##### - Windows: Open the Anaconda Prompt program. You should see a command prompt window.
##### - Mac/Linux: Your terminal window should already be open. Type command-t to open a new tab.

#### Type the following into the black window:
- Vader is the python library for processing textual data (another optional application is TextBlob).
> python -m pip install vaderSentiment
- we need to install some NLTK corpora. It helps to manage a large and structured set of texts.
> python -m textblob.download_corpora

Once complete, you can switch back to Jupytor notebooks.

In [1]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
analyser = SentimentIntensityAnalyzer()

In [2]:
analyser.polarity_scores("The movie is good")

{'neg': 0.0, 'neu': 0.508, 'pos': 0.492, 'compound': 0.4404}

The result states the following assumptions of the text:
- Negative ('neg') to be 0.0
- Neutral ('neu') to be 0.508
- Positive ('pos') to be 0.492

The most important value in the above result is the 'compound'. This is the score of the text.

The Compound score has a range of [-1, 1], being:
- [-1 to 0): negative,
- [0]: neutral
- (0 to +1]: positive

Returning to our analysis, the sentiment analysis of the text denotes it is more positive than neutral and definitely not negative.

Let's check the different degrees of sentiment:

In [3]:
text1 = 'The movie is very bad' 
text1_score = analyser.polarity_scores(text1)
print(' “The movie is very bad” ==> Compound: ',text1_score['compound'])

text2 = 'The movie is VERY BAD' 
text2_score = analyser.polarity_scores(text2)
print(' “The movie is VERY BAD” ==> Compound: ',text2_score['compound'])

text3 = 'The movie is VERY BAD!!!' 
text3_score = analyser.polarity_scores(text3)
print(' “The movie is VERY BAD!!!” ==> Compound: ',text3_score['compound'])


 “The movie is very bad” ==> Compound:  -0.5849
 “The movie is VERY BAD” ==> Compound:  -0.7398
 “The movie is VERY BAD!!!” ==> Compound:  -0.7984


Notice how the value of the compound becomes more and more negative when you include capital letters and exclamation marks!

Here's another example where sentiment analysis can be done on a text from a different language. 

Note 1: Vader application is built in English so you will need to translate the text into English.
For that, we will use Googletrans, a free and unlimited python library that implemented Google Translate API (for details, please refer to the API Documentation).

Note 2: In order for the translation to work, you will need to be connected to the internet.

To install Googletrans, you can go back to the command prompt or terminal window and type the following command: 
> python -m pip install googletrans


In [4]:
from googletrans import Translator
translator = Translator()

In [5]:
# Spanish: "My sister is the best"
text = translator.translate('mi hermana esla mejor').text
analyser.polarity_scores(text)

{'neg': 0.0, 'neu': 0.488, 'pos': 0.512, 'compound': 0.6369}

Here's a function that will run the text through, if the language is English it will automatically perform sentiment analysis, else it will translate and then perform the analysis.

In [6]:
def sentiment_analyzer_scores(text, engl=True):
    if engl:
        trans = text
    else:
        trans = translator.translate(text).text
    score = analyser.polarity_scores(trans)
    lb = score['compound']
    if lb >= 0.05:
        return 1
    elif (lb > -0.05) and (lb < 0.05):
        return 0
    else:
        return -1

In [7]:
# Portuguese: “The day is beautiful, with a lot of sun”
text = 'o dia esta lindo, com muito sol'
sentiment_analyzer_scores(text, False)

1

In [8]:
# French: “It is very horrible”
text = "No! c'est tres horrible"
sentiment_analyzer_scores(text, False)

-1

Great! At this point, we can analyze the sentiment behind text in practically any language!

Here are some practical ways you can use this information:

- Create a spatial graph on where your company is mentioned the most around the world
- Run sentiment analysis on tweets to see if the overall opinion of your company is positive or negative
- Create a social graphs of the most popular users that tweet about your company or product


### References:
- https://www.geeksforgeeks.org/twitter-sentiment-analysis-using-python/
- https://towardsdatascience.com/almost-real-time-twitter-sentiment-analysis-with-tweep-vader-f88ed5b93b1c